FormalConjectures/Wikipedia/EllipticCurveRank.lean

```diff
@@ -0,0 +1,205 @@
+@[category research solved, AMS 11 14]
+theorem _08375_le_density_rank_zero_one : 0.8375 ≤ atTop.liminf
+    fun H ↦ ({E ∈ heightLE H | E.rank = 0 ∨ E.rank = 1}.ncard / (heightLE H).ncard : ℝ) := by
+  sorry
+
+/-- Theorem 5 of [BS2013]:
+when elliptic curves over ℚ are ordered by height, a density of at least 20.62% have rank 0. -/
+@[category research solved, AMS 11 14]
+theorem _02062_le_density_rank_zero : 0.2062 ≤ atTop.liminf
+    fun H ↦ ({E ∈ heightLE H | E.rank = 0}.ncard / (heightLE H).ncard : ℝ) := by
+  sorry
+
+/-- From [PPVW2016], Section 3.1: "from the mid-1960s to the present,
+it seems that most experts conjectured unboundedness". -/
+@[category research open, AMS 11 14]
+theorem unbounded_rank_conjecture (n : ℕ) : ∃ E : RatEllipticCurve, n ≤ E.rank := by
+  sorry
+
+/-- From [PPVW2016], Section 8.2:
+"Our heuristic predicts (a) All but finitely many E ∈ ℰ satisfy rk E(ℚ) ≤ 21".
+In other words, there are only finitely many elliptic curves over ℚ (up to isomorphism)
+with rank greater than 21.
+Notice that this contradicts the previous conjecture. -/
+@[category research open, AMS 11 14]
+theorem finite_twentyone_lt_finrank : {E : RatEllipticCurve | 21 < E.rank}.Finite := by
+  sorry
+
+/-- [PPVW2016] 8.2(b). Note: ℰ_H should be ℰ_{≤H}, see Theorem 7.3.3. -/
+@[category research open, AMS 11 14]
+theorem rank_height_count_asymptotic (r : ℕ) (h₁ : 1 ≤ r) (h₂ : r ≤ 20) :
+    ∃ f : ℕ+ → ℝ, atTop.Tendsto f (𝓝 0) ∧
+      ∀ H : ℕ+, {E ∈ heightLE H | r ≤ E.rank}.ncard = (H : ℝ) ^ ((21 - r) / 24 + f H) := by
+  sorry
+
+/-- [PPVW2016] 8.2(c). -/
+@[category research open, AMS 11 14]
+theorem twentyone_le_rank_height_count_asymptotic :
+    ∃ f : ℕ+ → ℝ, atTop.Tendsto f (𝓝 0) ∧
+      ∀ H : ℕ+, {E ∈ heightLE H | 21 ≤ E.rank}.ncard ≤ (H : ℝ) ^ f H := by
+  sorry
+
+end RatEllipticCurve
+
+namespace WeierstrassCurve
+
+/-! See https://en.wikipedia.org/wiki/Rank_of_an_elliptic_curve#Largest_known_ranks -/
+
+/-- The elliptic curve over ℚ of rank at least 29 found by Elkies and Klagsbrun in 2024.
+It has rank exactly 29 assuming the generalized Riemann hypothesis. -/
+def elkiesKlagsbrun29 : WeierstrassCurve ℚ where
+  a₁ := 1
+  a₂ := 0
+  a₃ := 0
+  a₄ := -27006183241630922218434652145297453784768054621836357954737385
+  a₆ := 55258058551342376475736699591118191821521067032535079608372404779149413277716173425636721497
+
+/-- See https://mathoverflow.net/a/478050. -/
+@[category test, AMS 11 14]
+theorem Δ_elkiesKlagsbrun29 : elkiesKlagsbrun29.Δ =
+    -2 ^ 19 * 3 ^ 7 * 5 ^ 7 * 7 ^ 4 * 11 ^ 5 * 13 ^ 3 * 17 ^ 4 * 31 ^ 3 * 41 ^ 2 * 43 ^ 2 * 61 ^ 2 *
+    233 * 241 ^ 2 * 4139 * 678146849364709860535420504397393 *
+    159788990966780131363155786084695062643236502969 *
+    4402149008473369392540402625019227412319473055901 := by
+  rw [elkiesKlagsbrun29, Δ, b₂, b₄, b₆, b₈]; norm_num
+
+@[category test, AMS 11 14]
+instance : elkiesKlagsbrun29.IsElliptic where
+  isUnit := by rw [Δ_elkiesKlagsbrun29]; norm_num
+
+@[category test, AMS 11 14]
```

Comment:
`test` seems a little weird to me (or it the rank calculation particularly easy in that case?). Does this seem sensible?
```suggestion
@[category graduate, AMS 11 14]
```

Reply:
Good point; what I had in mind is that given the [list of points](https://web.math.pmf.unizg.hr/~duje/tors/rk29.html) one can show they are linearly independent using a computer algebra system (see [this](https://math.stackexchange.com/a/3022032/12932) or the [PARI/GP code](https://listserv.nodak.edu/cgi-bin/wa.exe?A2=NMBRTHRY;b9d018b1.2409&FT=&P=&H=&S=b) given by Elkies). However, finding the 29 points is a research problem, so maybe even `research solved` is appropriate.

Reply:
let's to with `research solved`